File: erpnext/budget/report/actual_line_items/actual_line_items.js
// Copyright (c) 2023, Extension and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Actual Line Items"] = {
    "filters": [
        {
            "fieldname": "project",
            "label": __("Project"),
            "fieldtype": "MultiSelectList",
            get_data: function (txt) {
                return frappe.db.get_link_options('Project', txt, {});
            },
            "width": "80",
        },
        {
            "fieldname": "wbs",
            "label": __("WBS"),
            "fieldtype": "MultiSelectList",
            get_data: function (txt) {
                return frappe.db.get_link_options('Work Breakdown Structure', txt, {});
            }
        },
        {
            "fieldname": "show_group_totals",
            "label": __("Show Group Totals"),
            "fieldtype": "Check",
            "hidden": 1
        },
        {
            "fieldname": "fiscal_year",
            "label": __("Fiscal Year"),
            "fieldtype": "Link",
            "options": "Fiscal Year"
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "depends_on": 'eval:doc.fiscal_year'
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "depends_on": 'eval:doc.fiscal_year'
        },
        {
            "fieldname": "voucher_type",
            "label": __("Voucher Type"),
            "fieldtype": "Select",
            "options": [" ","Purchase Receipt", "Purchase Invoice", "Stock Entry", "Expense Claim", "Journal Entry"],
            "default": " "
        },
        {
            "fieldname": "voucher_name",
            "label": __("Voucher Name"),
            "fieldtype": "Link",
            "options": () => frappe.query_report.get_filter_value("voucher_type"),
            "depends_on": 'eval:doc.voucher_type'
        },
        {
            "fieldname": "supplier",
            "label": __("Supplier"),
            "fieldtype": "MultiSelectList",
            get_data: function (txt) {
                return frappe.db.get_link_options('Supplier', txt, {});
            }
        },
        // {
        //     "fieldname": "ec_type",
        //     "label": __("Expense Claim Type"),
        //     "fieldtype": "Link",
        //     "options": "Expense Claim Type",
        //     // "depends_on": "eval:doc.voucher_type == 'Expense Claim'"
        // },
        // {
        //     "fieldname": "se_type",
        //     "label": __("Stock Entry Type"),
        //     "fieldtype": "Link",
        //     "options": "Stock Entry Type"
        // },
        {
            "fieldname": "item_code",
            "label": __("Item Code"),
            "fieldtype": "MultiSelectList",
            get_data: function (txt) {
                return frappe.db.get_link_options('Item', txt, {});
            }
        },
        {
            "fieldname": "item_group",
            "label": __("Item Group"),
            "fieldtype": "MultiSelectList",
            get_data: function (txt) {
                return frappe.db.get_link_options('Item Group', txt, {});
            }
        },
        // {
        //     "fieldname": "purchase_order",
        //     "label": __("Purchase Order"),
        //     "fieldtype": "MultiSelectList",
        //     get_data: function (txt) {
        //         return frappe.db.get_link_options('Purchase Order', txt, {});
        //     }
        // },
        // {
        //     "fieldname": "gl_account",
        //     "label": __("GL Account"),
        //     "fieldtype": "Link",
        //     "options": "Account"
        // },
        // {
        //     "fieldname": "cost_center",
        //     "label": __("Cost Center"),
        //     "fieldtype": "MultiSelectList",
        //     get_data: function (txt) {
        //         return frappe.db.get_link_options('Cost Center', txt, {});
        //     }
        // },
        // {
        //     "fieldname": "plant",
        //     "label": __("Plant"),
        //     "fieldtype": "MultiSelectList",
        //     get_data: function (txt) {
        //         return frappe.db.get_link_options('Plant', txt, {});
        //     }
        // },

    ],
    "formatter": function (value, row, column, data, default_formatter) {
    value = default_formatter(value, row, column, data);

    if (column.field === "WBS") {
        value = $(`<span>${value}</span>`);
        var $value = $(value).css("font-weight", "bold");
        value = $value.wrap("<p></p>").parent().html();
    }

    if (data && row[1] && row[1].content === "Grand Total") {
        value = $(`<span>${value}</span>`);
        var $value = $(value).css("font-weight", "bold");
        value = $value.wrap("<p></p>").parent().html();
    }

    return value;
}


};

File: erpnext/budget/report/actual_line_items/actual_line_items.py
import frappe

def execute(filters=None):
    if not filters:
        filters = {}

    columns = [
        {"label": "WBS", "fieldname": "wbs", "fieldtype": "Link", "options": "Work Breakdown Structure", "width": 200},
        {"label": "WBS Name", "fieldname": "wbs_name", "fieldtype": "Data", "width": 150},
        {"label": "WBS Level", "fieldname": "wbs_level", "fieldtype": "Int", "width": 100},
        {"label": "Voucher Type", "fieldname": "voucher_type", "fieldtype": "Data", "width": 100},
        {"label": "Voucher Name", "fieldname": "voucher_no", "fieldtype": "Data", "width": 150},
        {"label": "Voucher Date", "fieldname": "voucher_date", "fieldtype": "Date", "width": 100},
        {"label": "Item", "fieldname": "item", "fieldtype": "Data", "width": 100},
        {"label": "Qty", "fieldname": "qty", "fieldtype": "Float", "width": 150},
        {"label": "Rate", "fieldname": "rate", "fieldtype": "Float", "width": 100},
        {"label": "Amount", "fieldname": "amount", "fieldtype": "Float", "width": 100},
    ]

    conditions = []
    params = {}

    if filters.get("fiscal_year"):
        fiscal_year = filters["fiscal_year"]
        fiscal_year_data = frappe.db.get_value("Fiscal Year", fiscal_year, ["year_start_date", "year_end_date"], as_dict=True)

        if fiscal_year_data:
            from_date = fiscal_year_data["year_start_date"]
            to_date = fiscal_year_data["year_end_date"]
            conditions.append("be.posting_date BETWEEN %(from_date)s AND %(to_date)s")
            params["from_date"] = from_date
            params["to_date"] = to_date

    elif filters.get("from_date") and filters.get("to_date"):
        conditions.append("be.posting_date BETWEEN %(from_date)s AND %(to_date)s")
        params["from_date"] = filters["from_date"]
        params["to_date"] = filters["to_date"]

    if filters.get("project"):
        conditions.append("be.project IN %(project)s")
        params["project"] = tuple(filters["project"]) 

    if filters.get("wbs"):
        conditions.append("be.wbs IN %(wbs)s")
        params["wbs"] = tuple(filters["wbs"])

    if filters.get("voucher_type"):
        conditions.append("be.voucher_type = %(voucher_type)s")
        params["voucher_type"] = filters["voucher_type"]

    if filters.get("voucher_name"):
        conditions.append("be.voucher_no = %(voucher_name)s")
        params["voucher_name"] = filters["voucher_name"]

    if filters.get("supplier"):
        conditions.append("pi.supplier IN %(supplier)s")
        params["supplier"] = tuple(filters["supplier"])

    if filters.get("item_code"):
        conditions.append("pii.item_code IN %(item_code)s")
        params["item_code"] = tuple(filters["item_code"])

    if filters.get("item_group"):
        conditions.append("pii.item_group IN %(item_group)s")
        params["item_group"] = tuple(filters["item_group"])

    vouchers = ["Purchase Invoice", "Purchase Receipt"]
    conditions.append("be.voucher_type IN %(vouchers)s")
    params["vouchers"] = tuple(vouchers)

    base_query = """
        SELECT 
            be.wbs AS wbs,
            be.wbs_name AS wbs_name,
            be.wbs_level AS wbs_level,
            be.voucher_type AS voucher_type,
            be.voucher_no AS voucher_no,
            be.posting_date AS voucher_date,
            CASE
                WHEN be.voucher_type = 'Purchase Invoice' THEN pii.item_code
                WHEN be.voucher_type = 'Purchase Receipt' THEN mri.item_code
                ELSE NULL
            END AS item,
            CASE
                WHEN be.voucher_type = 'Purchase Invoice' THEN pii.qty
                WHEN be.voucher_type = 'Purchase Receipt' THEN mri.qty
                ELSE NULL
            END AS qty,
            CASE
                WHEN be.voucher_type = 'Purchase Invoice' THEN pii.rate
                WHEN be.voucher_type = 'Purchase Receipt' THEN mri.rate
                ELSE NULL
            END AS rate,
            CASE
                WHEN be.voucher_type = 'Purchase Invoice' THEN pii.qty * pii.rate
                WHEN be.voucher_type = 'Purchase Receipt' THEN mri.qty * mri.rate
                ELSE 0
            END AS amount
        FROM 
            `tabBudget Entry` AS be
        LEFT JOIN 
            `tabPurchase Invoice` AS pi ON pi.name = be.voucher_no
        LEFT JOIN 
            `tabPurchase Invoice Item` AS pii ON pii.parent = pi.name
        LEFT JOIN 
            `tabMaterial Request` AS mr ON mr.name = be.voucher_no
        LEFT JOIN 
            `tabMaterial Request Item` AS mri ON mri.parent = mr.name
    """

    if conditions:
        base_query += " WHERE " + " AND ".join(conditions)

    data = frappe.db.sql(base_query, params, as_dict=True)

    total_qty = sum(row.get("qty", 0) for row in data)
    total_amount = sum(row.get("amount", 0) for row in data)

    data.append({})
    data.append({
        "wbs": "Grand Total",
        "wbs_name": "",
        "wbs_level": None,
        "voucher_type": "",
        "voucher_no": "",
        "voucher_date": None,
        "item": "",
        "qty": total_qty,
        "rate": "",
        "amount": total_amount,
    })

    return columns, data
